src/diagonal.works/b6/ui/histogram.go

```diff
@@ -0,0 +1,248 @@
+		if n == nn {
+			ids = i
+			break
+		}
+	}
+	if ids < 0 {
+		ids = len(p.Ids)
+		p.Namespaces = append(p.Namespaces, n)
+		p.Ids = append(p.Ids, &pb.IDsProto{})
+	}
+	p.Ids[ids].Ids = append(p.Ids[ids].Ids, id.Value)
+}
+
+type bound struct {
+	lower, upper interface{}
+	label        string
+	within       func(interface{}) (bool, error)
+}
+type buckets []bound
+
+func xBound(lower, upper interface{}) bound {
+	return bound{
+		lower: lower,
+		upper: upper,
+		label: fmt.Sprint(lower) + "-" + fmt.Sprint(upper),
+		within: func(v interface{}) (bool, error) {
+			below, err := api.Less(lower, v)
+			if err != nil {
+				return false, err
+			}
+
+			above, err := api.Greater(upper, v)
+			if err != nil {
+				return false, err
+			}
+
+			return !below && !above, nil
+		},
+	}
+}
+
+func (b buckets) bucket(v interface{}) (int, error) {
+	for i, bound := range b {
+		yes, err := bound.within(v)
+		if err != nil {
+			return -1, err
+		}
+		if yes {
+			return i, nil
+		}
+	}
+
+	return -1, nil // Allowing data not to be bucketed.
+}
+
+const maxBuckets = 5
+
+func categorical(kvs []kv) (buckets, error) {
+	sort.Slice(kvs, func(i, j int) bool {
+		return kvs[i].value > kvs[j].value
+	})
+
+	var b buckets
+	for i, kv := range kvs {
+		if i == maxBuckets {
+			break
+		}
+
+		key := kv.key
+		if i == maxBuckets-1 && len(kvs) > maxBuckets {
+			key = "other"
+		}
+
+		b = append(b, bound{lower: key, label: fmt.Sprintf("%s", key), within: func(v interface{}) (bool, error) { return v == key || key == "other", nil }})
+	}
+
+	return b, nil
+}
+
+func uniform(kvs []kv) (buckets, error) {
+	sort.Slice(kvs, func(i, j int) bool {
+		greater, err := api.Greater(kvs[i].key, kvs[j].key)
+		if err != nil {
+			panic(err) // Not graceful, but greater should handle all numericals,
+		} // and we do the numerical check in histogram call.
+
+		return greater
+	})
+
+	var b buckets
+	if (len(kvs)) <= maxBuckets {
+		for _, kv := range kvs {
+			b = append(b, bound{lower: kv.key, label: fmt.Sprintf("%s", kv.key), within: func(v interface{}) (bool, error) { return v == kv.key, nil }})
+		}
+	} else {
+		for len(kvs) > 0 {
+			bucket_size := len(kvs) / (maxBuckets - len(b))
+			b = append(b, xBound(kvs[0].key, kvs[bucket_size-1].key))
+			kvs = kvs[bucket_size:]
+		}
+	}
+
+	for i := 0; i < len(b)/2; i++ {
+		b[i], b[len(b)-1-i] = b[len(b)-1-i], b[i]
+	}
+
+	return b, nil
+}
+
+type kv struct {
+	key      interface{}
+	value    int            // key-count.
+	features []b6.FeatureID // features in this bucket, if relevant
+}
+
+func countValues(c api.Collection) ([]kv, error) {
+	m := make(map[interface{}]*kv)
+
+	i := c.Begin()
+	for {
+		ok, err := i.Next()
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			break
+		}
+
+		var e *kv
+		if e, ok = m[i.Value()]; ok {
+			e.value++
+		} else {
+			e = &kv{key: i.Value(), value: 1}
+			m[i.Value()] = e
+		}
+		if id, ok := i.Key().(b6.Identifiable); ok {
+			e.features = append(e.features, id.FeatureID())
```

Comment:
i m not sure i understand y we move the code actually, we could expose feature ids from upper layer also no? it feels a bit weird to have a histogram fn that just wraps the collection / but whatever